### ISSUE TITLE
Gap-aware weight (ωU,D) removal improves performance on MuSiQue

### ISSUE THREAD

**Issue by lufeiyang598** [meaningful96/CoopRAG#2]
Hi, thanks for releasing the code.

I’m reproducing the retriever on MuSiQue using the official implementation (batch size = 30 due to memory limits). I noticed that removing the gap-aware weight ωU,D (Wq) — i.e., not multiplying logits by wq_all — consistently improves retrieval performance.

According to the paper’s ablation, removing ωU,D should hurt performance, so I wanted to ask:

Is this behavior on MuSiQue expected?

Is ωU,D sensitive to batch size or data characteristics?

Any insight would be appreciated. Thanks!

**Comment by meaningful96** [comment]
Hello. First of all, thank you for your interest in my research.

The MuSiQue dataset is among the most challenging multi-hop question answering datasets. This is because the dataset creation process involved graph-based evidence collection, which means that even when the correct documents are aggregated, the contextual connections between them can be weak and introduce noise. For this reason, even Large Language Models (LLMs) find the MuSiQue dataset challenging.

The weight ($\omega_{U,D}$) variance per sample is significantly larger in this dataset compared to others, and thus a higher number of samples (i.e., a larger batch size) helps ensure training stability. Specifically, when a sample-wise weight, such as the gap-aware loss applied on top of InfoNCE, is multiplied, high variance in the weight distribution can lead to unstable training. I also experienced this issue in my previous work ([SATKGC, WWW2025](https://arxiv.org/abs/2407.12703)). This is likely the problem that manifested when the batch size was reduced.

Therefore, to directly address your questions:

1.  **This result is, to some extent, an expected outcome** on the MuSiQue dataset, especially when the batch size is small.
2.  The weight $\omega_{U,D}$ is indeed **sensitive to batch size and data characteristics**.

Thank you again for your valuable interest and feedback.

**Comment by lufeiyang598** [comment]
Thanks for the explanation. I understand the sensitivity of ωU,D to batch size and data characteristics, especially on MuSiQue. That said, I’m still a bit puzzled that reducing the batch size from 40 to 30 could lead to such a drastic change in behavior. I’ll continue experimenting with larger batches and related settings to better understand this effect. Thanks again for the insights.

**Comment by lufeiyang598** [comment]
Thank you for your previous replies. I’d like to follow up with a more concise but direct question.

In my experiments, removing the gap-aware weight ω_{U,D} does not lead to the degradation reported in the paper:

on MuSiQue, retrieval performance consistently improves;

on HotpotQA, performance does not decrease and in some runs slightly improves.

This seems inconsistent with the paper’s ablation results and the conclusion that ω_{U,D} provides a stable performance gain.

Concretely, I remove ω_{U,D} in model.py by commenting out fixed_weight_query() and replacing the weight with ones, while keeping the rest of the scoring unchanged:

# Wq = self.fixed_weight_query(q_h, d_h, self.args.layers)
# pos_wq = Wq[:, 0::2]
# neg_wq = Wq[:, 1::2]
# wq_all = torch.cat([pos_wq, neg_wq], dim=1)

wq_all = torch.ones_like(logits)
logits_weighted = logits * wq_all


All other training and evaluation settings are identical to the official configuration.

Could you please help clarify why ω_{U,D} does not show the expected positive effect in these cases?
Are there specific conditions or assumptions under which the benefit of ω_{U,D} is expected to hold?

Thank you for your clarification and for making the code available.

**Comment by meaningful96** [comment]
Thank you for the careful repro. After investigating, I found small but impactful differences between my local training code and the version pushed to the repo. While trying to reduce memory usage, I had introduced precision/optimizer tweaks that changed the training dynamics; this likely explains why the gap-aware weight ​did not yield the same gains on your side.

There is another important discrepancy: in the paper runs the per-query batch size was larger. The number of negatives per query matches the current code, but we trained with 512 queries per batch, so the logits tensor was effectively 512 × 80. This behaves like gradient accumulation over many queries, stabilizing the score distribution; under that setting, the gap-aware weighting tends to provide a more consistent benefit.

I’ve confirmed the issue and will roll back to the original setup and update the repository. Your report was very helpful in pinpointing this. I’ll post an update once the fix is pushed. Thanks again for flagging this and for sharing concrete code changes and results.

**Comment by lufeiyang598** [comment]
Thanks for the detailed clarification — this explains the discrepancy very well.

I really appreciate you investigating this and updating the repo. Looking forward to trying the fix once it’s pushed. Thanks again!
